core/src/main/java/com/datastax/oss/driver/internal/core/config/map/DefaultOptionsMap.java

```diff
@@ -0,0 +1,189 @@
+/*
+ * Copyright DataStax, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.datastax.oss.driver.internal.core.config.map;
+
+import com.datastax.oss.driver.api.core.config.DriverOption;
+import com.datastax.oss.driver.api.core.config.map.OptionsMap;
+import com.datastax.oss.driver.api.core.config.map.TypedDriverOption;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import java.time.Duration;
+import java.time.temporal.ChronoUnit;
+import java.util.Collections;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class DefaultOptionsMap implements OptionsMap {
+
+  private final ConcurrentHashMap<String, Map<DriverOption, Object>> map =
+      new ConcurrentHashMap<>();
+
+  @Nullable
+  @Override
+  public <ValueT> ValueT put(
+      @NonNull String profile, @NonNull TypedDriverOption<ValueT> option, @NonNull ValueT value) {
+    Objects.requireNonNull(option, "option");
+    Objects.requireNonNull(value, "value");
+    Object previous = getProfileMap(profile).put(option.getRawOption(), value);
+    return cast(previous);
+  }
+
+  @Nullable
+  @Override
+  public <ValueT> ValueT get(@NonNull String profile, @NonNull TypedDriverOption<ValueT> option) {
+    Objects.requireNonNull(option, "option");
+    Object result = getProfileMap(profile).get(option.getRawOption());
+    return cast(result);
+  }
+
+  @Nullable
+  @Override
+  public <ValueT> ValueT remove(
+      @NonNull String profile, @NonNull TypedDriverOption<ValueT> option) {
+    Objects.requireNonNull(option, "option");
+    Object previous = getProfileMap(profile).remove(option.getRawOption());
+    return cast(previous);
+  }
+
+  @NonNull
+  @Override
+  public Map<String, Map<DriverOption, Object>> asRawMap() {
+    return map;
+  }
+
+  @NonNull
+  private Map<DriverOption, Object> getProfileMap(@NonNull String profile) {
+    Objects.requireNonNull(profile, "profile");
+    return map.computeIfAbsent(profile, p -> new ConcurrentHashMap<>());
+  }
+
+  // Isolate the suppressed warning for retrieval. The cast should always succeed unless the user
+  // messes with asMap() directly.
+  @SuppressWarnings({"unchecked", "TypeParameterUnusedInFormals"})
+  @Nullable
+  private <ValueT> ValueT cast(@Nullable Object value) {
```

Comment:
We might be able to improve things a bit:

```java
@Nullable
private <ValueT> ValueT cast(@NonNull GenericType<ValueT> expectedType, @Nullable Object value) {
  if (value == null) {
    return null;
  }
  if (expectedType.isSupertypeOf(GenericType.of(value.getClass()))) {
    @SuppressWarnings("unchecked")
    ValueT cast = (ValueT) value;
    return cast;
  }
  // The cast should always succeed unless the user messes with asMap() directly.
  throw new ClassCastException("Cannot cast blah blah");
}
```

Reply:
How is that better? As far as I can tell, it does exactly the same thing as the cast operator.